[Week3] 과제 구현

sonms opened the pull request:
## 📌 PR 요약

🫛 작업한 내용
- MVVM, AAC ViewModel 적용
- 카드 애니메이션
- [x] 필수 과제
- [x] 심화 과제

🫛 PR 포인트
- 애니메이션의 작동 상태
- 객체가 가져야 할 정보란?

## 📸 스크린샷
| 필수 | 심화 |
|:-----------:|:-----------:|
| <video width="360" src="https://github.com/user-attachments/assets/d164e994-81b2-4567-9c1e-85e56c74e4b1" /> |<video width="360" src="https://github.com/user-attachments/assets/4c264359-e806-40e5-a556-79bb9edfcba8" /> |

## 📮 관련 이슈
- closed #이슈번호

Commit:
feat: Coil 라이브러리 추가 및 인터넷 권한 설정

- 이미지 로딩을 위한 Coil 라이브러리 의존성을 `build.gradle.kts`와 `libs.versions.toml`에 추가했습니다.
- 네트워크 통신을 위해 `AndroidManifest.xml`에 `INTERNET` 권한을 추가했습니다.

Commit:
feat: 검색 화면 상태 관리 개선 및 UX 향상

- `SearchModel`에 `isFlipped` 상태를 추가하고, 상태를 토글하는 `flip()` 확장 함수를 정의했습니다.
- 검색 화면의 상태 관리를 `flippedIndices` Set에서 `SearchModel` 내부의 `isFlipped` 프로퍼티를 직접 사용하도록 변경했습니다.
- 아이템 클릭 시 인덱스가 아닌 고유 `id`를 사용하여 뒤집기 상태를 업데이트하도록 수정하여, 검색 결과가 필터링되어도 각 아이템의 뒤집기 상태가 유지되도록 개선했습니다.
- `LazyColumn`의 `key`로 아이템의 `id`를 지정하여 리스트 성능과 안정성을 향상시켰습니다.
- 검색어 입력 시 300ms의 디바운스(debounce)를 적용하여 불필요한 검색 실행을 줄였습니다.
- `DiveSoptItem` 컴포저블의 `onFlip` 파라미터 이름을 `onClick`으로 변경하여 의미를 명확히 했습니다.

Commit:
feat: 홈 화면 UI 모델 및 상태 정의

- `HomeImageUiModel` 데이터 클래스를 추가하여 홈 화면의 이미지 정보를 관리합니다. 이 모델은 ID, 앞면/뒷면 이미지 URL, 그리고 뒤집힘 상태(`isFlipped`)를 포함합니다.
- `HomeUiState` 데이터 클래스를 정의하여 홈 화면의 UI 상태를 관리합니다. 이 상태는 `HomeImageUiModel`의 `ImmutableList`를 포함합니다.

Commit:
feat: 홈 화면 ViewModel 구현

- 홈 화면의 UI 상태를 관리하는 `HomeViewModel`을 추가했습니다.
- `HomeUiState`를 정의하여 홈 화면의 이미지 데이터 목록을 관리합니다.
- `HomeImageUiModel`을 사용하여 각 이미지 카드의 ID, 앞/뒤 이미지 URL 및 뒤집힘 상태(`isFlipped`)를 표현합니다.
- `toggleFlip` 함수를 구현하여 특정 ID의 이미지 카드를 뒤집는 로직을 처리합니다.

Commit:
feat: 카드 뒤집기 애니메이션 컴포저블 구현

- `HomeImageHolder` 컴포저블을 새로 추가하여 카드 뒤집기 UI 및 애니메이션을 구현했습니다.
- `isFlipped` 상태에 따라 카드가 3초 동안 7번 회전하며 앞면과 뒷면을 전환하는 애니메이션을 적용했습니다.
- `LaunchedEffect`와 `Animatable`을 사용하여 회전 애니메이션을 관리합니다.
- `graphicsLayer`의 `rotationY`와 `cameraDistance`를 조정하여 3D 회전 효과를 구현했습니다.
- Coil 라이브러리의 `AsyncImage`를 사용하여 앞면과 뒷면의 이미지를 표시합니다.

Commit:
feat: 적용한 ViewModel의 데이터를 토대로 뷰 구현

Commit:
refactor: 사용하지 않는 import 문 제거

- `SearchScreen.kt`에서 더 이상 사용되지 않는 import 문들을 정리했습니다.

Commit:
fix: 카드 뒤집기 애니메이션 초기 렌더링 오류 수정

- `HomeImageHolder` 컴포저블에서 `isFlipped` 상태가 처음 설정될 때 불필요한 애니메이션이 발생하는 문제를 해결했습니다.
- `isInitial` 상태 변수를 추가하여, 컴포저블이 처음 렌더링될 때는 애니메이션 없이 즉시 카드의 앞/뒤 상태(`rotation`)를 설정하도록 수정했습니다.

Commit:
feat: 이스터에그 화면 추가 및 마이페이지에서 진입 로직 구현

- `ProfileImageHolder`를 클릭하면 이스터에그 화면으로 이동하도록 `clickable` 수정자를 추가했습니다.
- `MainAppState`에 `navigateToEasterEgg` 함수를 정의하고, `NavHost`에 `easterEggGraph`를 추가하여 이스터에그 화면으로의 네비게이션을 구현했습니다.
- 마이페이지에서 이스터에그 화면으로 이동할 수 있도록 관련 콜백 함수(`navigateEasterEgg`)를 `MyPageGraph`와 `MyPageRoute`에 전달했습니다.

Commit:
feat: 이스터에그 화면 네비게이션 설정

- 이스터에그 화면으로 이동하기 위한 네비게이션 그래프(`easterEggGraph`)와 확장 함수(`navigateEasterEgg`)를 추가했습니다.
- 네비게이션 라우트를 정의하는 `EasterEgg` 객체를 추가했습니다.
- 이스터에그에 사용될 이미지 리소스(`img_god_dongmin.png`)를 추가했습니다.

Commit:
feat: 이스터에그 화면 및 애니메이션 구현

- 클릭 시 이미지가 3D 회전하며 텍스트 카드의 뒤로 이동하는 애니메이션을 구현했습니다.
- `updateTransition`을 사용하여 클릭 상태(`isClicked`)에 따른 여러 애니메이션 값(offset, z-index, alpha)을 동시에 관리합니다.
- `Animatable`과 `LaunchedEffect`를 이용해 Y축 기준 360도 회전 애니메이션을 적용했습니다.
- `graphicsLayer`를 사용하여 이미지의 `translationX`, `translationY`, `rotationY`를 동적으로 변경합니다.
- 클릭 시 이미지와 텍스트 카드의 `zIndex`를 교체하여 앞뒤 순서를 변경하는 효과를 추가했습니다.

oilbeaneda commented on app/src/main/AndroidManifest.xml:
```diff
@@ -2,6 +2,8 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools">
 
+    <uses-permission android:name="android.permission.INTERNET" />
```
인터넷에 있는 이미지 사용하려면 이렇게 추가해주면되는거죠?! 이번 4차 셈나(맞나..)에서 배운것처럼요 !!

sonms replied:
맞습니다!~!

oilbeaneda commented on app/src/main/java/com/sopt/dive/presentation/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,72 @@
+
+    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()
+
+    fun toggleFlip(id: Int) {
```
이함수는 어떤 역할을 하는지 궁금합니다!!

sonms replied:
현재 아이템 들 중 어느 아이템이 뒤집혔는지 갱신하는 함수입니다!

oilbeaneda commented on app/src/main/java/com/sopt/dive/presentation/mypage/MyPageScreen.kt:
```diff
@@ -38,7 +27,8 @@ import kotlinx.coroutines.launch
 @Composable
 fun MyPageRoute(
```
MyPageRoute 함수를 MyPageScreen 파일 내부에 작성하신 이유가 있는지 궁금합니다!  제가 생각한 navcontroller 의 route 와는 다른 역할인지도 궁금합니두

sonms replied:
[https://github.com/37-dive-sopt-android/minseong-son/pull/3#discussion_r2443313016](https://github.com/37-dive-sopt-android/minseong-son/pull/3#discussion_r2443313016)  이 내용 + navcontroller 의 진입점이라고 생각하셔도 좋을 것 같습니다
 Screen 은 Ui를 그리는 책임만 가지기 위해 그리고 MyPage 화면을 그려주는게 최종 목적이니 Screen이라는 파일 이름 안에 작성했습니다~

oilbeaneda reviewed (commented):
팟짱님의 오른쪽  검지 손가락의 손톱 오비 손민성님 너무 수고많으셨습니다!!! 늘 자세한 설명과 코리 감사합니다🤩

hyeminililo commented on app/src/main/java/com/sopt/dive/presentation/easteregg/EasterEggScreen.kt:
```diff
@@ -0,0 +1,195 @@
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Dp
```
보통 이렇게 하나의 패키지에서 임포트 하는 양이 많을 때 `ui.*` 이런식으로 하는 경우가 있던데 따로 안 하신 이유가 있나요 ,,?

sonms replied:
우선 [https://developer.android.com/kotlin/style-guide?hl=ko](https://developer.android.com/kotlin/style-guide?hl=ko) 공식문서 기준으로 import 문에 *는 권장되지 않습니다!

> Avoid wildcard imports (*).
They make it unclear which classes are actually used and can lead to unexpected name clashes.

import androidx.compose.ui.*로 한 번에 가져오면 실제로 쓰지 않는 클래스·함수까지 다 import 후보로 포함돼요
그래서 사용되지 않는 import들이 많아지면 빌드 시 불필요한 탐색이 늘어나고 코드 가독성이 떨어지게 됩니다

```kotlin
//1번 방식
import androidx.compose.ui.*

//2ㅓ번 방식
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
```

두 방식 중에서는 아래 방식이 어떤 기능을 사용하고 있는지 확인하기 쉽죠? 이러한 이유로 인해 와일드 카드는 지양해야해요

hyeminililo commented on app/src/main/java/com/sopt/dive/presentation/home/component/HomeImageHolder.kt:
```diff
@@ -0,0 +1,96 @@
+        isFlipped = false,
+        onClick = {}
+    )
+}
```
아티클에 나와있는 것처럼  `EOL`이  뜨지 않도록 하면 좋을 것 같아요 ,,!

hyeminililo commented on app/src/main/java/com/sopt/dive/presentation/home/component/HomeImageHolder.kt:
```diff
@@ -0,0 +1,96 @@
+        modifier = modifier
+            .graphicsLayer {
+                rotationY = rotation.value % 360f
+                cameraDistance = 12 * density
```
이건 어떤 효과를 주는 코드인가요 ?? 그리고 이렇게 고정해서 사용하신 이유가 궁금합니다 !

sonms replied:
graphicsLayer { ... }은 해당 컴포저블의 그래픽 변환(translation, rotation, scale 등) 을 직접 지정할 수 있는 블록으로 현재는 rotationY 즉 Y 축으로 회전 시키고 카메라 거리로 입체감을 조절하고 있어요!

hyeminililo commented on app/src/main/java/com/sopt/dive/presentation/home/model/HomeImageUiModel.kt:
```diff
@@ -0,0 +1,11 @@
+import androidx.compose.runtime.Immutable
+
+@Immutable
+data class HomeImageUiModel(
```
나중에 이미지를 추가할 수 있도록 컴포넌트로 분리하는 방법 너무 좋은 것 같네요 ! 
그런데 `@Immutable` 어노테이션을 이용해 data class를 불변으로 만든 것 같은데 이게 어떤 걸 의미하나요 ?  싱글톤 패턴인 object를 활용하는 것과 다른건가요 ?

sonms replied:
immutable 은 클래스의 인스턴스는 생성 후 내부 상태가 불변한다는 것을 compose에게 알려주는 어노테이션으로 값이 바뀌지 않는 객체임을 컴파일러에게 보장하는 용도에요 그렇기 때문에 recomposition을 생략할 수 있어요 이 데이터는 변하지 않았으니까 다시 그릴 필요 없어라고 하는 거죠

object는 어플리케이션 전체에서 하나의 인스턴스만 존재하여 언제 어디서든 동일한 객체를 참조할 수 있도록 하는 공유 객체에요 그리고 만약 object에 var 변수가 존재한다면 그 값은 상태 바꾸기가 가능해서 불변이 아니겠죠? 둘은 이런 차이가 있어요

hyeminililo commented on app/src/main/java/com/sopt/dive/presentation/mypage/MyPageScreen.kt:
```diff
@@ -68,23 +58,28 @@ fun MyPageRoute(
 @Composable
 fun MyPageScreen(
     paddingValues: PaddingValues,
     userInfoList: List<Pair<String, String>>,
```
여기서 Pair로 관리하신 이유가 있으신가요 ??

sonms replied:
위에 userInfoList 를 사용하기 위해 선언했습니다~

hyeminililo reviewed (commented):
역시 오비님 ,, 항상 코드를 보면 깔끔하다는 생각이 드는 코드를작성하시는 것 같아요 ,,! 과제 너무 수고하셨습니다 !

sohee6989 commented on app/src/main/java/com/sopt/dive/presentation/home/component/HomeImageHolder.kt:
```diff
@@ -0,0 +1,96 @@
+            model = if (isBackVisible) backImageUrl else frontImageUrl,
+            contentDescription = null,
+            modifier = Modifier
+                .size(200.dp, 300.dp)
```
요거 파라미터명 명시해주면 더 가독성 좋을 것 같아요!!

sohee6989 commented on app/src/main/java/com/sopt/dive/core/designsystem/component/DiveSoptItem.kt:
```diff
@@ -69,9 +69,11 @@ fun DiveSoptItem(
             } else {
```
저번 코드여서 리뷰를 못 남겨서 여기다 남깁니다...!ㅎㅎ
이거 Box 안 써줘도 되지 않나요??

sonms replied:
추후 확장성이긴하는데 Card 안에 들어가는 Text을 감싸는 부모 컨테이너로 축 역할 + 나중에 image가 추가되어야할때 가운데 정렬 하려구 추가했습니다!

sohee6989 commented on app/src/main/java/com/sopt/dive/presentation/easteregg/EasterEggScreen.kt:
```diff
@@ -0,0 +1,195 @@
+                .noRippleClickable(onClick = { isClicked = !isClicked }),
+            contentAlignment = Alignment.Center
+        ) {
+            Box(
```
이 부분도 Box로 안 만들어도 될 것 같은데 Box를 쓰신 이유가 궁금합니다!

sonms replied:
시각적 컨테이너 역할이에요 물론 Text로 조절이 가능하다고 해도 Text자체가 너무 많은 책임을 지고 있다고 생각했어요 - 코너 조절, 색상, alpha, 등 그래서 Box를 통해 뒷면 구조를 구현했어요 !

sohee6989 commented on app/src/main/java/com/sopt/dive/presentation/easteregg/EasterEggScreen.kt:
```diff
@@ -0,0 +1,195 @@
+    var isInitial by remember { mutableStateOf(true) }
+
+    // isClicked 감지하여 새로운 목표 설정가능하게
+    val transition = updateTransition(targetState = isClicked, label = "image_transition")
```
우와 심화과제 완전 복잡하다..........

sohee6989 reviewed (commented):
수고하셨습니다!!
오늘도 코드 잘 보고 배워갑니다~!!

한 가지 궁금한게 있어서 코멘트에다 질문 남깁니다..!!
영상 올려주신거 보면 홈화면 애니메이션 구현해주신 것 중에 맨 마지막 아이템은 처음 돌아갈 때 뒷면 사진이 빈 상태로 나타나다가 멈췄을 때 뒤에 사진이 보이고 그 이후로 돌아갈 때는 돌아갈 때도 뒷면이 사진이 잘 보이고 있던데 이거 의도하신건가요????

sonms commented:
> 수고하셨습니다!! 오늘도 코드 잘 보고 배워갑니다~!!
> 
> 한 가지 궁금한게 있어서 코멘트에다 질문 남깁니다..!! 영상 올려주신거 보면 홈화면 애니메이션 구현해주신 것 중에 맨 마지막 아이템은 처음 돌아갈 때 뒷면 사진이 빈 상태로 나타나다가 멈췄을 때 뒤에 사진이 보이고 그 이후로 돌아갈 때는 돌아갈 때도 뒷면이 사진이 잘 보이고 있던데 이거 의도하신건가요????

아 사진이 보이는 거는 로딩이 된거고 맨아래는 로딩이 안되서 안보이는 겁니다 ㅠㅠ

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/home/component/HomeImageHolder.kt:
```diff
@@ -0,0 +1,96 @@
+    Card(
+        modifier = modifier
+            .graphicsLayer {
+                rotationY = rotation.value % 360f
```
최근에 드는 생각, 지향하고자 하는 스타일인데
'rotation.value % 360f'처럼 간단한 표현이라도 코드가 길어지거나 변경될 때 계속 해석해야 해서 생각보다 리소스를 쓰게 되더라구요.
그래서 이런 값은 wrapRotation() 같은 형태로 함수화해두면 네이밍만 보고 역할이 바로 보이고, 조건문 안에서 사용할 때도 훨씬 읽기 편해져서 요즘 선호하고 잇슴니다ㅎㅎ

sonms replied:
오!! 저도 적용해봐야겠네요 padding도 따로 빼서 쓰라는 말도 있던데 한 번 더 찾아봐야겠어요!  좋은 내용 감사합니다! 😘

Commit:
refactor: 코드리뷰 적용 - `wrapRotation` 확장 함수를 추가하여 회전 로직 개선

- `Float` 타입에 대한 `wrapRotation` 확장 함수를 추가하여, 회전 값을 360도로 나누는 나머지 연산을 표준화했습니다.
- `HomeImageHolder` 컴포저블에서 `rotationY`와 `isBackVisible`을 계산할 때 기존의 `% 360f` 연산을 새로 만든 `wrapRotation()` 함수로 대체했습니다.
